Add support for dnf

KaratekHD opened the pull request:
These changes should allow for dnf to be used as a backend instead of zypper.
I made it so that opi creates the file `~/.config/opi/config.json` with the following content on first start:
```json
{
  "backend": "zypp"
}
```
You can then change the value to `dnf` if you like to.
Please provide feedback for this so that I can improve this!
~Jens

asdil12 commented:
Thanks for the contribution - the fact that you switched the indentation of all files you touched makes it hard to see what you actually did. Could you switch it back to tab indentation?

KaratekHD commented:
I did? I probably have to check my IDE configuratin..
Sure, will change that

KaratekHD commented:
It's changed, but the diffs are still not really readable... If you want to, I can open a new pull request from a new fork without the messed up indents

asdil12 commented on opi/__init__.py:
```diff
@@ -265,9 +329,13 @@ def ask_number(min_num, max_num, question="Choose a number (0 to quit):"):
+
 def ask_keep_repo(repo):
 	if not ask_yes_or_no('Do you want to keep the repo "%s"?' % repo, 'y'):
-		subprocess.call(['sudo', 'zypper', 'rr', repo])
```
Beware that this might get called not with a repo alias (=filename) but with a repo name.

KaratekHD replied:
I'll work on these changes, but especially this gives me quite a headache and I am not yet sure how I can do this

asdil12 replied:
Maybe have a look where this is actually called - I tried to make my latest changes use the repo alias (=filename) but I'm not 100% sure if it is called everywhere with the alias. Maybe it would be easier to make sure that it is only called with the alias.

asdil12 replied:
Alternatively this could be handled by parsing all repo files and comparing the names.

asdil12 replied:
I just had a look and it seems that there is currently no remaining call with the repo name.
The only call that was using the repo name instead of the alias was moved to the alias in https://github.com/openSUSE/opi/commit/50c43d064468c87c34713ecb6e4f6ca8550bee89

asdil12 commented on opi/config/__init__.py:
```diff
@@ -0,0 +1,33 @@
+import pwd
+import subprocess
+import json
+from functools import cache
```
Have a look at the testsuite: https://github.com/openSUSE/opi/runs/2906061366?check_suite_focus=true

KaratekHD replied:
I think the test should work now, but I can not run them because I can not get podman to work.

asdil12 commented:
The PR that deduplicates the repo creation code is merged now, so it should be easier to build on top of that - you will need to rebase, though.

KaratekHD commented:
I hope this didn't mess anything up, this was the first rebase I ever did :sweat_smile:

KaratekHD commented:
Alright, I am happy with this now, should I also bump the version number up?

asdil12 reviewed (changes requested):
The code formatting was changed again making the diff pretty unreadable.
Please also squash your commits together where it makes sense.

asdil12 commented:
> Alright, I am happy with this now, should I also bump the version number up?

Not needed - this will be done on the next release.

KaratekHD commented:
> The code formatting was changed again making the diff pretty unreadable.
> Please also squash your commits together where it makes sense.
Changed it. I squashed some commits, but I can't figure out how to squash more of them, I hope this is enough

asdil12 commented:
There are still many unrelated code style changes (spaces and linebreaks) and it seems that this is not properly rebased. Eg. you readded old code to the `add_packman_repo` function.
I think it would be best to start freshly.
You could backup your work somewhere and `git rebase --hard origin master` to get a fresh working space.
To fix existing commits you can use `git commit --amend` to fix the last commit and `git commit --fixup=[hash_of_commit_to_be_fixed] ; git rebase -i --autosquash [hash_of_commit_to_be_fixed]~1` if there are commits in between.
With `git rebase -i master` you can reorganize your commits, squash them, reorder them, etc.

KaratekHD commented:
I will reopen this soon.

KaratekHD commented:
Alright, I reworked it, please have a look at it.

asdil12 commented on opi/__init__.py:
```diff
@@ -72,13 +73,16 @@ def add_packman_repo(dup=False):
 
 
-############
+############i
```
```suggestion
############
```

asdil12 commented on opi/__init__.py:
```diff
@@ -101,16 +105,25 @@ def add_repo(filename, name, url, enabled=True, gpgcheck=True, gpgkey=None, repo
+			refresh_cmd.append('--gpg-auto-import-keys')
+		refresh_cmd.append('ref')
+	elif get_backend() == BackendConstants.dnf:
+		refresh_cmd = ['sudo', 'dnf', 'ref']
```
Will this automatically import the gpg keys as well?

KaratekHD replied:
According to https://unix.stackexchange.com/questions/424543/how-to-import-key-that-signs-a-repository-for-dnf-yum it shares the key database with RPM, so `rpm --import` should be enough, and it is called before the refresh happens.

asdil12 commented on opi/__init__.py:
```diff
@@ -248,7 +261,12 @@ def install_binary(binary):
+		if get_backend() == BackendConstants.zypp:
+			flags = ['--allow-vendor-change', '--allow-arch-change', '--allow-downgrade', '--allow-name-change']
+		elif get_backend() == BackendConstants.dnf:
+			flags = ['--setopt=allow_vendor_change=True'] # TODO: architecture changes, name changes, downgrades?
```
Please make sure to find alternatives to these flags from zypper.
Also it might make sense to not use backend specific flag as args to the `install_packages` function anymore but normal kwargs like `allow_vendor_change=False`. This way all the backend specifics could be handled in `install_packages` instead of in `install_packages`.

KaratekHD replied:
Moving the flags to `install_packages()` is a good idea, will do that. `--allow-downgrade` and `--allow-name-change` are, according to @Conan-Kudo, automatically active in dnf. For arch changes, we need to specify the target architecture, so we need to somehow figure that out.

asdil12 commented on opi/__init__.py:
```diff
@@ -101,18 +105,41 @@ def add_repo(filename, name, url, enabled=True, gpgcheck=True, gpgkey=None, repo
-	args = ['sudo', 'zypper', 'in']
-	if from_repo:
-		args.extend(['--from', from_repo])
+def install_packages(packages, from_repo=None, flags=None, allow_vendor_change=False, allow_arch_change=False, allow_downgrade=False, allow_name_change=False):
+	if get_backend() == BackendConstants.zypp:
+		args = ['sudo', 'zypper', 'in']
+		if from_repo:
+			args.extend(['--from', from_repo])
+	elif get_backend() == BackendConstants.dnf:
+		args = ['sudo', 'dnf', 'in']
+		if from_repo:
+			args.extend(['--repo', from_repo])
 	if flags:
 		args.extend(flags)
```
I think you can remove the `flags` argument as well as the processing of it, as we don't use it anymore.

asdil12 commented on opi/__init__.py:
```diff
@@ -101,18 +105,41 @@ def add_repo(filename, name, url, enabled=True, gpgcheck=True, gpgkey=None, repo
 		args.extend(flags)
+	if get_backend() == BackendConstants.zypp:
+		if allow_downgrade:
+			args.extend(['--allow-downgrade'])
```
If you only add a single element, better use `append`:

```suggestion
			args.append('--allow-downgrade')
```
